torch/csrc/jit/codegen/cuda/kernel_cache.cpp

```diff
@@ -294,10 +136,51 @@ InputsIdLookup::IdLookupReturn InputsIdLookup::lookupId(
 FusionExecutorCache::FusionExecutorCache(std::unique_ptr<Fusion> fusion)
     : fusion_(std::move(fusion)) {}
 
+// Note [ Channels Last support in nvfuser ]
+//
+// Background:
+// To support channels last in nvfuser with optimal performance, we would want
+// to allow dimension collapsing in generated code on channels-last tensors,
+// which greatly simplifies indexing. Current API in codegen only allows
+// dimensional collapsing on neighboring axes. The unfortunate thing is that
+// memory format design in PyTorch is implicitly marked by strides, while the
+// semantics meaning of axes remain unchanged. i.e. A 4d tensor with axes [N, C,
+// H, W] would have the same shape in both format, while contiguous tensor
+// carries strides [C*H*W, H*W, W, 1] and channels-last tensor [H*W*C, 1, W*C,
+// C].
+//
+// Approach:
+// Part_1. To allow axes collapsing for channels-last format in codegen, we can
+// permute input tensor to have axes in decending order by their strides, so
+// they would be viewed as `contiguous` in codegen, hence collapsed to simple
+// indexing. Part_2. To ensure correct result, we need to ensure computation in
+// nvfuser carries same semantics as with TorchScript graph. We need to
+//   Part_2_1. Maintain a bookkeeping where each codegen tensor is tagged with
+//   either `contiguous` format or `channels_last` format. Part_2_2. Parsing
+//   rule should handle and propagate the tag properly, i.e. having special
+//   rules for `channels_last` input tensor and mark output in its right format.
+// Part_3. Codegen output tensor in `channels_last` format should be permuted
+// back to `contiguous` format before returning to TorchScript
+//
+// For details  on Part_2, refer to implementation Note [ Format Bookkeeping and
+// Propagation in Parser ]
 std::vector<at::Tensor> FusionExecutorCache::runFusionWithInputs(
-    const at::ArrayRef<IValue>& inputs) {
+    const at::ArrayRef<IValue>& inputs_) {
   FUSER_PERF_SCOPE("FusionExecutorCache::runFusionWithInputs");
 
+  // permute input tensor for kernel execution. See Part_1 in Note [ Channels
+  // Last support in nvfuser ]
+  at::ArrayRef<IValue> inputs = inputs_;
```

Comment:
Nit: maybe the `inputs_` parameters should be left as `inputs` and the local variable be changed to `perm_inputs` so as not to confuse the input parameter as a data member?